File: tools/lib/zeo_dev/commands/bench.rb
# frozen_string_literal: true

require "fileutils"
require "tmpdir"

module ZeoDev
  module Commands
    # The COMPILER-cost instrument: what a compile costs and produces per
    # program (frontend wall time, emitted CLIF lines, peak RSS, binary
    # size) across a fixed hello-world-to-bundler ladder.
    #
    # The RUNTIME performance suite lives in criterion now:
    #
    #   cargo bench -p zeo --bench programs
    #
    # (see bench/README.md for baselines, comparisons, and profiling).
    class Bench < Cli
      # The fixed program set: names paired with repo-relative paths,
      # spanning hello-world to bundler scale. `hello` is synthesized so the
      # smallest possible program is always in the record.
      COMPILE_PROGRAMS = [
        ["hello", nil],
        ["bm_fib", "bench/bm_fib.rb"],
        ["bm_template", "bench/bm_template.rb"],
        ["bm_json_parse", "bench/bm_json_parse.rb"],
        ["bm_micro_lisp", "bench/bm_micro_lisp.rb"],
        ["core_classes", "tests/core_classes.rb"],
        ["uri_parse_and_build", "tests/uri_parse_and_build.rb"],
        ["optparse_subset", "tests/optparse_subset.rb"],
        ["gem_rubygems", "tests/bench/rubygems.rb"],
        ["gem_bundler", "tests/bench/bundler.rb"]
      ].freeze

      COMPILE_BASELINE = "bench/compile-baseline.tsv"

      def self.summary = "the compiler-cost instrument (runtime suite: cargo bench)"

      def self.banner
        "usage: zeo-dev bench --compile [--filter <substr>] [--runs N] [--update-baseline]"
      end

      def defaults = { filter: nil, runs: 3, update: false, compile: false }

      def options(o)
        o.on("--filter SUBSTR", "only programs whose name contains SUBSTR") { |v| opts[:filter] = v }
        o.on("--runs N", Integer, "compiles per program (default 3)") { |v| opts[:runs] = v }
        o.on("--update-baseline", "bank the measured rows into the baseline") { opts[:update] = true }
        o.on("--compile", "measure the compiler (the only mode)") { opts[:compile] = true }
      end

      def run
        expect_no_args!
        raise Error, "bench: --runs takes a positive integer" unless opts[:runs].positive?
        unless opts[:compile]
          raise Error, "bench: the runtime suite moved to criterion -- " \
                       "run `cargo bench -p zeo --bench programs` " \
                       "(this command keeps only --compile)"
        end

        @zeo = Ruby.build_zeo!
        run_compile
      end

      private

      def path(rel) = File.join(ROOT, rel)

      def selected(name) = opts[:filter].nil? || name.include?(opts[:filter])

      # A progress line the caller sees IMMEDIATELY, even piped or
      # backgrounded: stdout through a pipe is block-buffered.
      def progress(line)
        puts line
        $stdout.flush
      end

      # `frontend_ms` is best-of-N `--emit-clif` wall time, the renderer the
      # BUILD path uses. `peak_rss` takes the WORST run rather than the best:
      # time is best-of-N because the fastest run is the least disturbed, but
      # memory is a ceiling question, and the most a compile ever held is the
      # number that decides whether it fits.
      def run_compile
        rows = {}
        COMPILE_PROGRAMS.each do |name, rel|
          next unless selected(name)

          rb = rel ? path(rel) : write_hello
          begin
            rows[name] = compile_one(name, rb)
            r = rows[name]
            progress(format("%-22s %6dms  %6d lines  %5d MiB  %8d bytes",
                            name, r[:frontend_ms], r[:lines], r[:peak_rss] / (1024 * 1024),
                            r[:bin_bytes]))
          rescue Error => e
            progress(format("%-22s FAIL  %s", name, e.message))
            return 1
          end
        end
        return puts("no programs matched") || 1 if rows.empty?

        if opts[:update]
          write_compile_baseline(rows)
          puts "baseline updated: #{path(COMPILE_BASELINE)}"
        end
        0
      end

      def write_hello
        p = File.join(Dir.tmpdir, "zeo-bench-hello.rb")
        File.write(p, %(puts "hello"\n))
        p
      end

      def compile_one(name, rb)
        emitted = File.join(Dir.tmpdir, "zeo-bench-#{name}.clif")
        best_ms = nil
        lines = 0
        peak = 0
        opts[:runs].times do
          t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
          res = Exec.run([@zeo, rb, "--emit-clif=#{emitted}", "-W0"], env: { "ZEO_TIMINGS" => "1" })
          ms = ((Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0) * 1000).round
          raise Error, "zeo --emit-clif failed: #{res.stderr.lines.first&.strip}" unless res.success?

          lines = timing_field(res.stderr, "lines=") || 0
          peak = [peak, timing_field(res.stderr, "peak_rss=") || 0].max
          best_ms = best_ms.nil? ? ms : [best_ms, ms].min
        end
        FileUtils.rm_f(emitted)

        # One `-o` for the binary size. Nothing caches a built binary, so
        # this measures the real cost.
        bin = File.join(Dir.tmpdir, "zeo-bench-bin-#{name}")
        res = Exec.run([@zeo, rb, "-o", bin, "-W0"])
        raise Error, "zeo -o failed: #{res.stderr.lines.reject { |l| l.start_with?("zeo-timings:") }.first&.strip}" unless res.success?

        bytes = File.size(bin)
        FileUtils.rm_f(bin)
        { frontend_ms: best_ms, lines: lines, peak_rss: peak, bin_bytes: bytes }
      end

      # One `key=value` field off the last `zeo-timings:` line carrying it.
      def timing_field(stderr, key)
        stderr.lines.reverse_each do |l|
          next unless l.start_with?("zeo-timings:")

          tok = l.split.find { |t| t.start_with?(key) }
          next unless tok

          return tok.delete_prefix(key).delete_suffix("ms").to_i
        end
        nil
      end

      def write_compile_baseline(rows)
        out = +"# bench/compile-baseline.tsv -- zeo-dev bench --compile --update-baseline\n"
        out << "# name\tfrontend_ms\tlines\tpeak_rss\tbin_bytes\n"
        rows.keys.sort.each do |n|
          r = rows[n]
          out << Tsv.row(n, r[:frontend_ms], r[:lines], r[:peak_rss], r[:bin_bytes]) << "\n"
        end
        Tsv.write(path(COMPILE_BASELINE), out)
      end
    end
  end
end

File: crates/xtask/src/commands/bench.rs
//! The COMPILER-cost instrument: what a compile costs and produces per
//! program (frontend wall time, emitted CLIF lines, peak RSS, binary size)
//! across a fixed hello-world-to-bundler ladder.
//!
//! The RUNTIME performance suite lives in criterion:
//!
//! ```text
//! cargo bench -p zeo --bench programs
//! ```
//!
//! (see bench/README.md for baselines, comparisons, and profiling).

use std::io::Write;
use std::path::{Path, PathBuf};
use std::time::Instant;

use crate::exec::{self, Capture};
use crate::scratch::Scratch;
use crate::{Error, root, root_join, write_if_changed};

/// The fixed program set: names paired with repo-relative paths, spanning
/// hello-world to bundler scale. `hello` is synthesized so the smallest
/// possible program is always in the record.
const PROGRAMS: &[(&str, Option<&str>)] = &[
    ("hello", None),
    ("bm_fib", Some("test/bench/bm_fib.rb")),
    ("bm_template", Some("test/bench/bm_template.rb")),
    ("bm_json_parse", Some("test/bench/bm_json_parse.rb")),
    ("bm_micro_lisp", Some("test/bench/bm_micro_lisp.rb")),
    ("core_classes", Some("test/lang/classes/core_classes.rb")),
    ("uri_parse_and_build", Some("test/stdlib/uri/uri_parse_and_build.rb")),
    ("optparse_subset", Some("test/stdlib/optparse/optparse_subset.rb")),
    ("gem_rubygems", Some("test/bench/compile/rubygems.rb")),
    ("gem_bundler", Some("test/bench/compile/bundler.rb")),
];

const BASELINE: &str = "test/bench/compile-baseline.tsv";

const USAGE: &str =
    "usage: cargo xtask bench --compile [--filter <substr>] [--runs N] [--update-baseline]";

struct Row {
    frontend_ms: u128,
    lines: u64,
    peak_rss: u64,
    bin_bytes: u64,
}

pub fn run(args: &[String]) -> Result<(), Error> {
    let mut filter = None;
    let mut runs: usize = 3;
    let mut update = false;
    let mut compile = false;
    let mut rest = args.iter();
    while let Some(arg) = rest.next() {
        match arg.as_str() {
            "--filter" => {
                filter = Some(
                    rest.next()
                        .cloned()
                        .ok_or_else(|| Error::new(format!("--filter wants a value\n{USAGE}")))?,
                );
            }
            "--runs" => {
                let value = rest
                    .next()
                    .ok_or_else(|| Error::new(format!("--runs wants a value\n{USAGE}")))?;
                runs = value
                    .parse()
                    .map_err(|_| Error::new("--runs takes a positive integer"))?;
            }
            "--update-baseline" => update = true,
            "--compile" => compile = true,
            "--help" | "-h" => {
                println!("{USAGE}");
                return Ok(());
            }
            other => return Err(Error::new(format!("unknown option {other:?}\n{USAGE}"))),
        }
    }
    if runs == 0 {
        return Err(Error::new("--runs takes a positive integer"));
    }
    if !compile {
        return Err(Error::new(
            "the runtime suite moved to criterion -- run `cargo bench -p zeo --bench programs` \
             (this command keeps only --compile)",
        ));
    }

    let zeo = crate::build_zeo()?;
    let work = Scratch::new("bench")?;
    let mut rows: Vec<(&str, Row)> = Vec::new();
    for (name, rel) in PROGRAMS {
        if filter.as_ref().is_some_and(|f| !name.contains(f.as_str())) {
            continue;
        }
        let rb = match rel {
            Some(rel) => root_join(rel),
            None => write_hello(work.path())?,
        };
        let row = match compile_one(&zeo, name, &rb, runs, work.path()) {
            Ok(row) => row,
            Err(e) => {
                progress(&format!("{name:<22} FAIL  {e}"));
                return Err(Error::reported());
            }
        };
        progress(&format!(
            "{name:<22} {:>6}ms  {:>6} lines  {:>5} MiB  {:>8} bytes",
            row.frontend_ms,
            row.lines,
            row.peak_rss / (1024 * 1024),
            row.bin_bytes
        ));
        rows.push((name, row));
    }
    if rows.is_empty() {
        println!("no programs matched");
        return Err(Error::reported());
    }
    if update {
        write_baseline(&rows)?;
        println!("baseline updated: {}", root_join(BASELINE).display());
    }
    Ok(())
}

/// A progress line the caller sees IMMEDIATELY, even piped or backgrounded:
/// stdout through a pipe is block-buffered.
fn progress(line: &str) {
    println!("{line}");
    let _ = std::io::stdout().flush();
}

fn write_hello(work: &Path) -> Result<PathBuf, Error> {
    let path = work.join("hello.rb");
    std::fs::write(&path, "puts \"hello\"\n")
        .map_err(|e| Error::new(format!("writing {}: {e}", path.display())))?;
    Ok(path)
}

/// `frontend_ms` is best-of-N `--emit-clif` wall time, the renderer the BUILD
/// path uses. `peak_rss` takes the WORST run rather than the best: time is
/// best-of-N because the fastest run is the least disturbed, but memory is a
/// ceiling question, and the most a compile ever held is the number that
/// decides whether it fits.
fn compile_one(zeo: &Path, name: &str, rb: &Path, runs: usize, work: &Path) -> Result<Row, Error> {
    let emitted = work.join(format!("{name}.clif"));
    let mut best_ms = None;
    let mut lines = 0;
    let mut peak = 0;
    for _ in 0..runs {
        let started = Instant::now();
        // Options BEFORE the file. zeo follows ruby's own convention, where
        // everything after the script name is the PROGRAM's ARGV -- so
        // `zeo prog.rb --emit-clif=x` runs the program and hands it the flag,
        // which is how this instrument came to time a full run and report it
        // as a compile.
        let emit = format!("--emit-clif={}", emitted.display());
        let out = exec::run(
            &[zeo, Path::new(&emit), Path::new("-W0"), rb],
            root(),
            &[("ZEO_TIMINGS", Some("1"))],
            Capture::Both,
        )?;
        let ms = started.elapsed().as_millis();
        if !out.success() {
            return Err(Error::new(format!(
                "zeo --emit-clif failed: {}",
                out.stderr_text().lines().next().unwrap_or_default().trim()
            )));
        }
        lines = timing_field(&out.stderr_text(), "lines=").unwrap_or(0);
        peak = peak.max(timing_field(&out.stderr_text(), "peak_rss=").unwrap_or(0));
        best_ms = Some(best_ms.map_or(ms, |b: u128| b.min(ms)));
    }

    // One `-o` for the binary size. Nothing caches a built binary, so this
    // measures the real cost.
    let bin = work.join(format!("bin-{name}"));
    let out = exec::run(
        &[zeo, Path::new("-o"), &bin, Path::new("-W0"), rb],
        root(),
        &[],
        Capture::Both,
    )?;
    if !out.success() {
        let text = out.stderr_text();
        let first = text
            .lines()
            .find(|l| !l.starts_with("zeo-timings:"))
            .unwrap_or_default();
        return Err(Error::new(format!("zeo -o failed: {}", first.trim())));
    }
    let bin_bytes = std::fs::metadata(&bin)
        .map_err(|e| Error::new(format!("reading {}: {e}", bin.display())))?
        .len();
    Ok(Row {
        frontend_ms: best_ms.unwrap_or(0),
        lines,
        peak_rss: peak,
        bin_bytes,
    })
}

/// One `key=value` field off the LAST `zeo-timings:` line carrying it.
fn timing_field(stderr: &str, key: &str) -> Option<u64> {
    for line in stderr.lines().rev() {
        if !line.starts_with("zeo-timings:") {
            continue;
        }
        let Some(token) = line.split_whitespace().find(|t| t.starts_with(key)) else {
            continue;
        };
        return token
            .trim_start_matches(key)
            .trim_end_matches("ms")
            .parse()
            .ok();
    }
    None
}

fn write_baseline(rows: &[(&str, Row)]) -> Result<(), Error> {
    let mut sorted: Vec<&(&str, Row)> = rows.iter().collect();
    sorted.sort_by_key(|(name, _)| *name);
    let mut out = String::from(
        "# bench/compile-baseline.tsv -- cargo xtask bench --compile --update-baseline\n\
         # name\tfrontend_ms\tlines\tpeak_rss\tbin_bytes\n",
    );
    for (name, r) in sorted {
        out.push_str(&format!(
            "{name}\t{}\t{}\t{}\t{}\n",
            r.frontend_ms, r.lines, r.peak_rss, r.bin_bytes
        ));
    }
    write_if_changed(&root_join(BASELINE), out.as_bytes())
}
